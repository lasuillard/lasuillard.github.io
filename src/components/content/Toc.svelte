<script lang="ts">
	import TocTree from '$components/content/TocTree.svelte';
	import { makeToc, type TreeNode, type TreeSource } from '$lib/toc';
	import { onMount } from 'svelte';

	export let content: HTMLElement | undefined;
	let rootHeadings: TreeNode<HTMLElement>[] = [];

	onMount(() => {
		if (!content) {
			console.error('Content element not available.');
			return;
		}

		const headings = [...content.querySelectorAll('h1, h2, h3, h4, h5, h6')].map(
			(elem) => elem as HTMLElement
		);
		const root = {
			data: null,
			children: [] as TreeNode<HTMLElement>[]
		};
		const items: TreeSource<HTMLElement, HTMLElement>[] = headings.map((v) => ({
			data: v,
			children: [],
			compare(other: TreeSource<HTMLElement, HTMLElement>): number {
				const left = this.data.tagName;
				const right = other.data.tagName;
				return left < right ? -1 : left == right ? 0 : 1;
			},
			toNode() {
				return {
					data: this.data,
					children: []
				};
			}
		}));
		makeToc(root, items);
		rootHeadings = root.children;
	});
</script>

<div data-testid="toc" {...$$restProps}>
	<div>
		{#each rootHeadings as root}
			<TocTree tree={root} />
		{/each}
	</div>
</div>
