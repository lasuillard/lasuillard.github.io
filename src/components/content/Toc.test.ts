// @vitest-environment happy-dom
import Toc from '$components/content/Toc.svelte';
import { render } from '@testing-library/svelte';
import { expect, it } from 'vitest';

it('has a valid locator', () => {
	const { getByTestId } = render(Toc);
	expect(getByTestId('toc')).toBeTruthy();
});
