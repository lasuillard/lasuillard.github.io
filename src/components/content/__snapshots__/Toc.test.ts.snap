// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`make ToC from given HTML element 1`] = `"<div><div data-testid=\\"toc-tree\\"><p class=\\"mb-1.5 font-light\\"><a class=\\"link-hover link\\">0</a></p> <ul></ul></div><!--<TocTree>--><div data-testid=\\"toc-tree\\"><p class=\\"mb-1.5 font-light\\"><a class=\\"link-hover link\\">1</a></p> <ul><li class=\\"ml-8\\"><div data-testid=\\"toc-tree\\"><p class=\\"mb-1.5 font-light\\"><a class=\\"link-hover link\\">2</a></p> <ul></ul></div><!--<TocTree>--> </li><li class=\\"ml-8\\"><div data-testid=\\"toc-tree\\"><p class=\\"mb-1.5 font-light\\"><a class=\\"link-hover link\\">3</a></p> <ul></ul></div><!--<TocTree>--> </li><li class=\\"ml-8\\"><div data-testid=\\"toc-tree\\"><p class=\\"mb-1.5 font-light\\"><a class=\\"link-hover link\\">4</a></p> <ul><li class=\\"ml-8\\"><div data-testid=\\"toc-tree\\"><p class=\\"mb-1.5 font-light\\"><a class=\\"link-hover link\\">5</a></p> <ul></ul></div><!--<TocTree>--> </li></ul></div><!--<TocTree>--> </li><li class=\\"ml-8\\"><div data-testid=\\"toc-tree\\"><p class=\\"mb-1.5 font-light\\"><a class=\\"link-hover link\\">6</a></p> <ul></ul></div><!--<TocTree>--> </li></ul></div><!--<TocTree>--><div data-testid=\\"toc-tree\\"><p class=\\"mb-1.5 font-light\\"><a class=\\"link-hover link\\">7</a></p> <ul><li class=\\"ml-8\\"><div data-testid=\\"toc-tree\\"><p class=\\"mb-1.5 font-light\\"><a class=\\"link-hover link\\">8</a></p> <ul><li class=\\"ml-8\\"><div data-testid=\\"toc-tree\\"><p class=\\"mb-1.5 font-light\\"><a class=\\"link-hover link\\">9</a></p> <ul><li class=\\"ml-8\\"><div data-testid=\\"toc-tree\\"><p class=\\"mb-1.5 font-light\\"><a class=\\"link-hover link\\">10</a></p> <ul></ul></div><!--<TocTree>--> </li></ul></div><!--<TocTree>--> </li></ul></div><!--<TocTree>--> </li><li class=\\"ml-8\\"><div data-testid=\\"toc-tree\\"><p class=\\"mb-1.5 font-light\\"><a class=\\"link-hover link\\">11</a></p> <ul><li class=\\"ml-8\\"><div data-testid=\\"toc-tree\\"><p class=\\"mb-1.5 font-light\\"><a class=\\"link-hover link\\">12</a></p> <ul></ul></div><!--<TocTree>--> </li></ul></div><!--<TocTree>--> </li></ul></div><!--<TocTree>--></div>"`;
